usautobuild/config.py

```diff
@@ -1,127 +1,171 @@
 import json
 import os
+import typing
 
+from collections.abc import Mapping
 from logging import getLogger
 from pathlib import Path
-from typing import Any, Optional
+from typing import Any, Optional, TypeVar, Union
 
-from usautobuild.exceptions import MissingRequiredEnv
-
-from .exceptions import InvalidConfigFile, MissingConfigFile
+from .exceptions import InvalidConfigFile
 
 log = getLogger("usautobuild")
 
 
+_UNSET = object()
+
+
+class Variable:
+    __slots__ = (
+        "default",
+        "type_",
+        "arg",
+        "config",
+        "env",
+    )
+
+    def __init__(
+        self,
+        default: Any = _UNSET,
+        *,
+        type_: Optional[Any] = None,
+        arg: Optional[str] = None,
+        config: Optional[str] = None,
+        env: Optional[str] = None,
+    ):
+        self.default = default
+        self.type_ = type_
+        self.arg = arg
+        self.config = config
+        self.env = env
+
+    def resolve(self, name: str, type_: Optional[Any], args: Mapping[str, Any], cfg: dict[str, Any]) -> Any:
+        if type_ is None:
+            if self.default is _UNSET:
+                raise ValueError(f"Need type annotation for {name} as there is no default value")
+
+            type_ = type(self.default)
+
+        value = self.default
+
+        if (env := self.env) is None:
+            env = name.upper()
+
+        # special handling for env syntax, make sure we are not touching default
+        if env in os.environ:
+            value = os.environ[env]
+
+            if type_ is bool:
+                if value.lower() in ("0", "no", "off", "disable", "false"):
+                    value = False
+                else:
+                    value = True
+            elif type_ is list:
+                value = value.split(",")
+
+        if (config := self.config) is None:
+            config = name
+
+        value = cfg.get(config, value)
+        value = args.get(config, value)
```

Comment:
there is a bug: should be `get(arg, value)`, same as with config but it isn't used anywhere now anyway

Reply:
fixed